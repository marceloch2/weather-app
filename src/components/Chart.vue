<template>
  <div class="weather__chart">
    <div v-for="(_type, i) in types" :key="i" class="weather__chart-tool">
      <button :disabled="type === _type" @click="changeType(_type)">{{formatTitle(_type)}}</button>
    </div>

    <div class="weather__chart-content">
      <div class="weather__chart_data">
        <img
          v-if="currentWeatherIcon"
          :src=" `http://openweathermap.org/img/w/${currentWeatherIcon}.png`"
        >
        <h1>{{ formatTitle(type) }}</h1>
        <h3>{{ currentWeatherDate }}</h3>
        <h4>{{ currentWeatherTime }}</h4>
        <h2>
          <span class="metric" v-html="metric"></span>
          {{ currentWeatherData }}
        </h2>
        <h5>{{ currentLocation.city }}</h5>
      </div>

      <div class="chart-container" style="position: relative; height: 10vh; width: 60vw">
        <canvas id="weather__chart-ref"></canvas>
      </div>
    </div>
  </div>
</template>

<script>
import { mapState } from "vuex";
import Chart from "chart.js";
import weatherApiService from "../Services/weatherApiService";

const weatherLineChart = null;

export default {
  data() {
    return {
      currentWeatherIcon: "",
      currentWeatherData: "",
      currentWeatherDate: "",
      currentWeatherTime: "",
      types: ["temperature", "wind_speed", "humidity"],
      type: "temperature",
      chartData: [],
      currentData: {
        temperature: [],
        humidity: [],
        wind_speed: [],
        labels: []
      }
    };
  },

  mounted() {
    this.resetData();
    this.renderChart();
    this.setData(true);
  },

  computed: mapState({
    currentLocation: state => state.location,
    weatherData: state => state.weatherData_forecast,
    metric: state => state.metric
  }),

  watch: {
    weatherData: function(_new, _old) {
      this.resetData();
      this.setData(true);
    }
  },

  methods: {
    formatTitle(title) {
      title = title.replace("_", " ");

      return title.charAt(0).toUpperCase() + title.slice(1);
    },

    changeType(type) {
      this.type = type;

      this.weatherLineChart.data.datasets[0].data = this.currentData[type];
      this.weatherLineChart.data.datasets[0].label = this.formatTitle(
        this.type
      );

      this.weatherLineChart.update();
    },

    async setData(updating = false) {
      let weatherResponse = this.weatherData;

      if (!weatherResponse) {
        throw new Error("Failed to load the weather app");
      } else if (!Object.keys(weatherResponse).length) return;

      weatherResponse.list.forEach((data, i) => {
        if (i === 0) {
          let currentID = data.weather[0];
          this.currentWeatherIcon = currentID.icon;
        }
        this.currentData.labels.push(data.dt_txt);
        this.currentData.temperature.push(data.main.temp);
        this.currentData.humidity.push(data.main.humidity);
        this.currentData.wind_speed.push(data.wind.speed);
      });

      this.weatherLineChart.data.datasets[0].data = this.currentData[this.type];
      this.weatherLineChart.data.datasets[0].label = this.formatTitle(
        this.type
      );

      this.weatherLineChart.update();
    },

    resetData() {
      this.currentData.labels = [];
      this.currentData.temperature = [];
      this.currentData.humidity = [];
      this.currentData.wind_speed = [];
      this.currentWeatherData = 0;
      this.currentWeatherIcon = "";
      this.currentWeatherData = "";
      this.currentWeatherDate = "";
      this.currentWeatherTime = "";
    },

    renderChart(chartData) {
      let _self = this,
        _currentDate;
      const ctx = document
        .getElementById("weather__chart-ref")
        .getContext("2d");

      Chart.defaults.global.defaultFontColor = "black";
      Chart.defaults.global.defaultFontSize = 16;

      const data = {
        labels: this.currentData.labels,
        datasets: [
          {
            label: this.formatTitle(this.type),
            fill: true,
            showTooltips: false,
            lineTension: 0.1,
            backgroundColor: "rgba(0, 100, 110, 0.8)",
            borderColor: "rgb(167, 105, 0)",
            borderCapStyle: "butt",
            borderDash: [],
            borderDashOffset: 0.0,
            borderJoinStyle: "miter",
            pointBorderColor: "white",
            pointBackgroundColor: "black",
            pointBorderWidth: 1,
            pointHoverRadius: 8,
            pointHoverBackgroundColor: "brown",
            pointHoverBorderColor: "yellow",
            pointHoverBorderWidth: 2,
            pointRadius: 5,
            pointHitRadius: 10,
            data: chartData || [],
            spanGaps: false
          }
        ]
      };

      var options = {
        responsive: true,
        // maintainAspectRatio: false,
        tooltips: {
          displayColors: false,
          enabled: false,

          custom: tooltipModel => {
            if (tooltipModel.body) {
              const titleLines = tooltipModel.title || [];
              const bodyLines = tooltipModel.body.map(data => data.lines);

              if (titleLines[0]) {
                _currentDate = new Date(titleLines[0]);
                _self.currentWeatherData = parseFloat(
                  bodyLines[0][0].split(":")[1]
                ).toFixed(2);
                _self.currentWeatherDate = _currentDate.toDateString();
                _self.currentWeatherTime = _currentDate.toTimeString();
              }
            }
          }
        },
        scales: {
          yAxes: [
            {
              ticks: {
                beginAtZero: true
              },
              scaleLabel: {
                display: true,
                labelString: "fahrenheit",
                fontSize: 12
              }
            }
          ]
        }
      };

      this.weatherLineChart = new Chart(ctx, {
        type: "line",
        data,
        options
      });
    }
  }
};
</script>

<style lang="scss" scoped>
.weather__chart {
  .weather__chart-content {
    width: 100%;
    display: flex;
    flex-flow: row;

    .weather__chart_data {
      width: 200px;
    }
  }

  .weather__chart-tool {
    display: inline;
  }

  h1 {
    text-align: center;
  }

  h2 {
    background-color: rgba(0, 100, 110, 0.8);
    height: 30px;
    color: #fff;
    padding: 10px;
  }

  button {
    margin: 5px;
  }
}
</style>