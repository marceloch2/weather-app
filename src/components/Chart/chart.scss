.weather__chart {
  display: flex;
  flex-flow: column;
  align-items: center;

  .weather__chart-content {
    width: 100%;
    display: flex;
    flex-flow: row;

    .weather__chart-info {
      width: 200px;

      p {
        margin: 0;
        padding: 5px;
      }

      .weather__chart-city {
        font-size: 2em;
        text-transform: uppercase;
      }

      .weather__chart-img {
        width: 100px;
      }

      .weather__chart-type {
        font-size: 1.6em;
      }

      .weather__chart-data {
        font-size: 3em;
      }
    }
  }

  .weather__chart-tool-wrapper {
    border: solid 1px rgba(0, 100, 110, 0.8);
    border-radius: 20px;
    margin: 10px;

    .weather__chart-tool {
      display: inline;
      padding: 10px;

      button {
        border: solid 1px rgba(0, 100, 110, 0.8);
        border-radius: 20px;
        padding: 5px 30px 5px 30px;
        cursor: pointer;
      }

      .active {
        background-color: rgba(0, 100, 110, 0.8);
        color: #fff;
      }
    }
  }

  h1 {
    text-align: center;
  }

  h2 {
    background-color: rgba(0, 100, 110, 0.8);
    height: 30px;
    color: #fff;
    padding: 10px;
  }

  button {
    margin: 5px;
  }
}
