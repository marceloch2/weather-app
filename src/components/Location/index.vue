<template>
  <form class="location" ref="locationForm">
    <fieldset>
      <legend>{{ title.toUpperCase() }}</legend>

      <label for="today__city">City Name</label>
      <input
        autofocus
        type="text"
        id="today__city"
        class="today__city"
        placeholder="city name"
        v-model.trim="city"
      >

      <label v-if="hideCountry" for="today__country">Country Code</label>
      <input
        v-if="hideCountry"
        type="text"
        id="today__country"
        class="today__country"
        placeholder="country code"
        v-model.trim="country"
      >
      <button type="submit">Update Data</button>
    </fieldset>
  </form>
</template>

<script>
import Location from "./location";
export default Location;
</script>

<style lang="sass" scoped>
	@import 'location.scss'
</style>
