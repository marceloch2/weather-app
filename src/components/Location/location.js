import { mapState } from 'vuex'

import weatherApiService from '../../Services/weatherApiService'

export default {
  name: 'today',
  data() {
    return {
      title: 'Choose your location',
      city: '',
      country: ''
    }
  },

  mounted() {},

  computed: mapState({
    location: state => state.location,
    tab: state => state.tab
  }),

  watch: {
    tab: function(_old, _new) {
      if (this.location.city !== '') {
        this.search()
      }
    }
  },

  methods: {
    async search() {
      let tab = this.$store.state.tab

      this.$store.commit('setCurrentLocation', {
        city: this.city,
        country: this.country
      })

      try {
        let weatherData = await weatherApiService.fetchWeather(
          this.$store.state.location.city,
          this.$store.state.location.country,
          tab
        )
        this.$store.commit('setWeatherData', {
          weatherData,
          tab
        })
      } catch (error) {
        this.$store.commit(
          'globalError',
          `No Weather data found for ${this.$store.state.location.city}`
        )
      }
    }
  }
}
