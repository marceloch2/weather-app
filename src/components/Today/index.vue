<template>
  <div class="today" v-if="weatherData.weather && !hasError && location.city">
    <img
      class="today__img"
      :src=" `http://openweathermap.org/img/w/${weatherData.weather[0].icon}.png`"
    >

    <p class="today__condition">{{ weatherData.weather[0].description }}</p>

    <p class="today__temp">
      <span class="metric" v-html="metric"></span>
      <strong>{{ weatherData.main.temp }}</strong>
    </p>

    <p class="today__wind">
      Wind speed:
      <strong>{{ weatherData.wind.speed }}</strong>
      <span class="metric" v-html="wind_speed"></span>
    </p>

    <p class="today__humidity">
      Humidity:
      <strong>{{ weatherData.main.humidity }}</strong>
      <span class="metric" v-html="humidity"></span>
    </p>

    <p class="today__city">City: {{ location.city }}</p>
  </div>
</template>

<script>
import Today from "./today";
export default Today;
</script>

<style lang="sass" scoped>
	@import 'today.scss'
</style>
