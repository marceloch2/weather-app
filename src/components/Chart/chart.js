import { mapState } from 'vuex'
import Chart from 'chart.js'
import weatherApiService from '../../Services/weatherApiService'

const weatherLineChart = null

export default {
  data() {
    return {
      activeClass: 'active',
      currentWeatherIcon: '',
      currentWeatherData: '',
      currentWeatherDate: '',
      currentWeatherTime: '',
      types: [ 'temperature', 'wind_speed', 'humidity' ],
      type: 'temperature',
      chartData: [],
      currentData: {
        temperature: [],
        humidity: [],
        wind_speed: [],
        labels: []
      }
    }
  },

  mounted() {
    this.resetData()
    this.renderChart()
    this.setData(true)
  },

  computed: mapState({
    currentLocation: state => state.location,
    weatherData: state => state.weatherData_forecast,
    metric: state => state.metric,
    wind_speed: state => state.wind_speed,
    humidity: state => state.humidity
  }),

  watch: {
    weatherData: function(_new, _old) {
      this.resetData()
      this.setData(true)
    }
  },

  methods: {
    formatTitle(title) {
      title = title.replace('_', ' ')

      return title.charAt(0).toUpperCase() + title.slice(1)
    },

    changeType(type) {
      this.currentWeatherData = ''
      this.type = type

      this.weatherLineChart.data.datasets[0].data = this.currentData[type]
      this.weatherLineChart.data.datasets[0].label = this.formatTitle(this.type)

      this.weatherLineChart.update()
    },

    async setData(updating = false) {
      let weatherResponse = this.weatherData

      if (!weatherResponse) {
        throw new Error('Failed to load the weather app')
      } else if (!Object.keys(weatherResponse).length) return

      weatherResponse.list.forEach((data, i) => {
        if (i === 0) {
          let currentID = data.weather[0]
          this.currentWeatherIcon = currentID.icon
        }
        this.currentData.labels.push(data.dt_txt)
        this.currentData.temperature.push(data.main.temp)
        this.currentData.humidity.push(data.main.humidity)
        this.currentData.wind_speed.push(data.wind.speed)
      })

      this.weatherLineChart.data.datasets[0].data = this.currentData[this.type]
      this.weatherLineChart.data.datasets[0].label = this.formatTitle(this.type)

      this.weatherLineChart.update()
    },

    resetData() {
      this.currentData.labels = []
      this.currentData.temperature = []
      this.currentData.humidity = []
      this.currentData.wind_speed = []
      this.currentWeatherData = 0
      this.currentWeatherIcon = ''
      this.currentWeatherData = ''
      this.currentWeatherDate = ''
      this.currentWeatherTime = ''
    },

    renderChart(chartData) {
      let _self = this,
        _currentDate
      const ctx = document.getElementById('weather__chart-ref').getContext('2d')

      Chart.defaults.global.defaultFontColor = 'black'
      Chart.defaults.global.defaultFontSize = 16

      const data = {
        labels: this.currentData.labels,
        datasets: [
          {
            label: this.formatTitle(this.type),
            fill: true,
            showTooltips: false,
            lineTension: 0.1,
            backgroundColor: 'rgba(0, 100, 110, 0.8)',
            borderColor: 'rgb(167, 105, 0)',
            borderCapStyle: 'butt',
            borderDash: [],
            borderDashOffset: 0.0,
            borderJoinStyle: 'miter',
            pointBorderColor: 'white',
            pointBackgroundColor: 'black',
            pointBorderWidth: 1,
            pointHoverRadius: 8,
            pointHoverBackgroundColor: 'brown',
            pointHoverBorderColor: 'yellow',
            pointHoverBorderWidth: 2,
            pointRadius: 5,
            pointHitRadius: 10,
            data: chartData || [],
            spanGaps: false
          }
        ]
      }

      var options = {
        responsive: true,
        tooltips: {
          displayColors: false,
          enabled: false,

          custom: tooltipModel => {
            if (tooltipModel.body) {
              const titleLines = tooltipModel.title || []
              const bodyLines = tooltipModel.body.map(data => data.lines)

              if (titleLines[0]) {
                _currentDate = new Date(titleLines[0])
                _self.currentWeatherData = parseFloat(bodyLines[0][0].split(':')[1]).toFixed(2)
                _self.currentWeatherDate = _currentDate.toDateString()
                _self.currentWeatherTime = _currentDate.toTimeString()
              }
            }
          }
        },
        scales: {
          yAxes: [
            {
              ticks: {
                beginAtZero: true
              },
              scaleLabel: {
                display: true,
                labelString: 'fahrenheit',
                fontSize: 12
              }
            }
          ]
        }
      }

      this.weatherLineChart = new Chart(ctx, {
        type: 'line',
        data,
        options
      })
    }
  }
}
