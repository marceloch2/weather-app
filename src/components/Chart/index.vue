<template>
  <div class="weather__chart">
    <div class="weather__chart-tool-wrapper">
      <div v-for="(_type, i) in types" :key="i" class="weather__chart-tool">
        <button
          v-bind:class="[_type === type ? activeClass : '']"
          :disabled="type === _type"
          @click="changeType(_type)"
        >{{formatTitle(_type)}}</button>
      </div>
    </div>

    <div class="weather__chart-content">
      <div class="weather__chart-info">
        <img
          class="weather__chart-img"
          v-if="currentWeatherIcon"
          :src=" `http://openweathermap.org/img/w/${currentWeatherIcon}.png`"
        >
        <p class="weather__chart-data" v-if="currentWeatherData">
          <span v-if="type === 'temperature'" class="metric" v-html="metric"></span>
          <span name="weather__chart-data-value">{{ currentWeatherData }}</span>
          <span v-if="type === 'wind_speed'" class="metric" v-html="wind_speed"></span>
          <span v-if="type === 'humidity'" class="metric" v-html="humidity"></span>
        </p>

        <p class="weather__chart-city">{{ currentLocation.city }}</p>
        <p class="weather__chart-type">{{ formatTitle(type) }}</p>
        <p class="weather__chart-date">{{ currentWeatherDate }}</p>
        <p class="weather__chart-time">{{ currentWeatherTime }}</p>
      </div>

      <div class="chart-container" style="position: relative; height: 10vh; width: 60vw">
        <canvas id="weather__chart-ref"></canvas>
      </div>
    </div>
  </div>
</template>

<script>
import Chart from "./chart";
export default Chart;
</script>

<style lang="scss" scoped>
@import "chart.scss";
</style>