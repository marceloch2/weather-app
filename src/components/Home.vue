<template>
  <div>
    <div class="app__topbar"></div>
    <div class="app__menu">
      <ul>
        <li>
          <a name="todayButton" href="#" @click.prevent="tabs('weather')">Today</a>
        </li>
        <li>
          <a name="chartButton" href="#" @click.prevent="tabs('forecast')">History</a>
        </li>
      </ul>
    </div>
    <div class="home">
      <div class="home__center">
        <Location :hasCountry="activeTab === 'weather' ? false : true "/>
        <div class="home__center-right">
          <div v-if="isLoading" class="lds-ripple">
            <div></div>
            <div></div>
          </div>

          <div v-if="!isLoading && !hasError" class="home__center-right-holder">
            <Chart v-show="activeTab === 'forecast'"/>
            <Today v-show="activeTab === 'weather'"/>
          </div>

          <div v-if="hasError && !isLoading" class="hasError">
            <span>:(</span>
          </div>
        </div>
      </div>

      <button class="logout" @click="logout">sign out</button>
      <span v-if="errorMessage" class="errorMessage">{{ errorMessage }}</span>
    </div>
  </div>
</template>

<script>
import { mapState } from "vuex";

import Location from "./Location";
import Today from "./Today";
import Chart from "./Chart";

import weatherApiService from "../Services/weatherApiService";
import { gmapApi } from "vue2-google-maps";

export default {
  data() {
    return {
      activeTab: "weather"
    };
  },
  methods: {
    /**
     *
     */
    async tabs(tab) {
      this.$store.commit("setTabType", tab);
      this.activeTab = tab;
    },

    /**
     *
     */
    logout() {
      this.$session.destroy();
      this.$router.push("/");
    }
  },

  computed: {
    google: gmapApi,
    isLoading() {
      return this.$store.state.isLoading;
    },
    errorMessage() {
      return this.$store.state.globalError.msg;
    },
    currentLocation() {
      return this.$store.state.location;
    },
    hasError() {
      return this.$store.state.globalError.hasError;
    }
  },

  mounted() {
    this.$store.commit("setCurrentLocation", {
      city: "",
      country: ""
    });
    this.$store.commit("globalError", "");
  },

  components: {
    Location,
    Chart,
    Today
  }
};
</script>

<style lang="scss">
@import "Global.scss";
</style>
