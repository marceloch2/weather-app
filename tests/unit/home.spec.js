import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import Home from '@/components/Home.vue'
import Chart from '@/components/Chart'
import Location from '@/components/Location'
import Today from '@/components/Today'

import createStateConfig from './createStateConfig'

const localVue = createLocalVue()
localVue.use(Vuex)

const stateConfig = createStateConfig()

const store = new Vuex.Store({
  state: stateConfig,
  mutations: {
    setCurrentLocation: jest.fn(),
    globalError: jest.fn(),
    setTabType: jest.fn()
  }
})

const $router = {
  push: jest.fn()
}

const $session = {
  destroy: () => {
    window.localStorage.removeItem('vue-session-key')
  },
  start: () => jest.fn()
}

const wrapper = shallowMount(Home, {
  store,
  localVue,
  mocks: {
    $router,
    $session
  }
})

describe('Home.vue', () => {
  afterEach(() => {
    window.localStorage.clear()
  })

  it('Snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('To have Location component', () => {
    expect(wrapper.find(Location).exists()).toBe(true)
  })

  it('To have Chart component', () => {
    expect(wrapper.find(Chart).exists()).toBe(true)
  })

  it('To have Today component', () => {
    expect(wrapper.find(Today).exists()).toBe(true)
  })

  it('Mount with initial state', () => {
    expect(wrapper.vm.$store.state.globalError.msg).toBe('')
    expect(wrapper.vm.$store.state.globalError.hasError).toBe(false)
    expect(wrapper.vm.$store.state.location.city).toBe('')
    expect(wrapper.vm.$store.state.location.country).toBe('')
  })

  it('Start with correct UI', () => {
    expect(wrapper.find('.app__menu').isVisible()).toBe(true)
    expect(wrapper.find('.app__topbar').isVisible()).toBe(true)
    expect(wrapper.find('.home').isVisible()).toBe(true)
    expect(wrapper.find('.home__center').isVisible()).toBe(true)
    expect(wrapper.find('.home__center-right').isVisible()).toBe(true)
  })

  it('Can call Tabs method and change local state', () => {
    wrapper.vm.tabs('forecast')
    expect(wrapper.vm.activeTab).toMatch('forecast')
  })

  it('Can logout', () => {
    window.localStorage.setItem(
      'vue-session-key',
      JSON.stringify({
        jwt: 'e96f86f596aa58253a2f77eef125ec5e'
      })
    )

    const session = () => JSON.parse(window.localStorage.getItem('vue-session-key'))

    expect(session().jwt).toMatch('e96f86f596aa58253a2f77eef125ec5e')

    wrapper.vm.logout()

    expect(session()).toBeNull()
  })

  it('Can isLoading works', () => {
    expect(wrapper.html()).toContain('<div class="home__center-right-holder">')
    expect(wrapper.find('.home__center-right-holder').isVisible()).toBe(true)

    wrapper.vm.$store.state.isLoading = true
    expect(wrapper.html()).not.toContain('<div class="home__center-right-holder">')
  })

  it('Can Error toast works', () => {
    wrapper.vm.$store.state.globalError.hasError = true
    wrapper.vm.$store.state.isLoading = false
    expect(wrapper.html()).toContain('<div class="hasError">')
    expect(wrapper.find('.hasError span').text()).toMatch(':(')

    wrapper.vm.$store.state.globalError.hasError = false
    expect(wrapper.html()).not.toContain('<div class="hasError">')
  })
})
