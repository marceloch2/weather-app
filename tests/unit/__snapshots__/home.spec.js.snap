// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home.vue Snapshot 1`] = `
<div>
  <div class="app__topbar"></div>
  <div class="app__menu">
    <ul>
      <li><a name="todayButton" href="#">Today</a></li>
      <li><a name="chartButton" href="#">History</a></li>
    </ul>
  </div>
  <div class="home">
    <div class="home__center">
      <location-stub></location-stub>
      <div class="home__center-right">
        <!---->
        <div class="home__center-right-holder">
          <chart-stub style="display: none;"></chart-stub>
          <today-stub></today-stub>
        </div>
        <!---->
      </div>
    </div> <button class="logout">sign out</button>
    <!---->
  </div>
</div>
`;
